docs/specs/mlnet-database-loader/mlnet-database-loader-specs.md

```diff
@@ -0,0 +1,407 @@
+
+Within the 'databases scope' problem there are multiple areas.
+
+The **scope** for this feature is initially limited to **relational databases** with higher priority on SQL Server and Azure SQL Database, but one of the goals is to make this loader/connector compatible with any relational database which is supported by .NET providers. 
+
+- Scope to support in this feature:
+    - Relational Databases, such as:
+        - SQL Server and Azure SQL Database
+        - Other relational databases such as MySQL, Oracle, PostgreSQL, etc.
+
+- Scope not supported in this feature:
+    - No-SQL Databases, such as:
+        - Azure Cosmos DB
+        - MongoDB
+        - Any other non-SQL database
+    - Big Data, such as:
+        - Big Data: Spark, Hadoop, etc.
+
+
+## Evidence
+
+Sample feedback and quotes:
+
+- *"[Rachel Poulsen comment](https://www.quora.com/What-are-the-most-valuable-skills-to-learn-for-a-data-scientist-now), Data Scientist at Silicon Valley Data Science: I have been a Data Scientist in Silicon Valley for the last 4 years. There are A LOT of tools out there for data science as I'm sure you are aware. But very importantly, SQL: I'd argue this is 100% necessary for all data scientists, regardless of whether you are using structured or unstructured data. Too many people and companies still use it and are comfortable with it for it to be thrown out with NoSQL techniques. Even companies using NoSQL databases and unstructured data are still using SQL."*
+
+- *"forki commented on May 17, 2018
+I did not say EF ;-) - I heard some of the authors of this package worked on EF before, but please please don't bind it to EF, or sql server or any other specific tech. Or course you want to get data into the learning system without going over to csv. But the data is usually already flattened anyway, so no need for a complex OR mapping tool - especially since the target data structure is Immutable"*
+
+- *Other customers - TBD*
+
+# Who is the Customer
+
+- .NET developers used to work with databases who are getting started with machine learning with ML.NET.
+
+
+# Goals
+
+The business goals are the following, depending on the possible scenarios:
+
+- Ability for developers to load and automatically stream data from relational databases in order to train/evaluate ML.NET models.
+- The code to load from a database should be extremely easy, a single line of code in most cases.
+- Tooling (Model Builder in VS and the CLI) and AutoML API should also support this feature.  
+
+# Solution
+
+The solution is to create an ML.NET database loader classes supporting the above scenarios. 
+
+The main supported features are:
+
+- **Easily load data into IDataView from databases**: The main goal is to load data into IDataView from databases very easily (a single line of code as one of the approaches) and properly (with good performance when streaming data while training).
```

Comment:
"Load" suggests that the data is physically copied from the DBMS to the machine running the training. While this is undoubtedly a common scenario, is it envisioned to be the only one?

Reply:
For this feature spec that the scenario in scope since it is probably the most common scenario:
Training-machine <--network--> Database-in-RDBMS-server

However, note that usually data will be streamed as the training process requests more data, in a similar way we do from large files in ML.NET. "Load" doesn't mean we load the whole database table or query 'in-memory' in the training machine unless the pipeline specifies to use cache, which should be used only for small datasets.

That's the most common scenario, for instance having an Azure SQL Database in the cloud or accessing data to any relational database (not only SQL Server) such as PostgreSQL, Oracle, MySQL, etc.

**Do you have any other scenario in mind?** 

Probably running the ML.NET code within SQL Server you mean?: [SQL Server CLR integration](https://docs.microsoft.com/en-us/sql/relational-databases/clr-integration/clr-integration-overview?view=sql-server-2017) - 
That scenario is also very interesting when training a model since the data is closer to the training process. 

However, that is initially out of the scope of this spec doc since it'll be a particular scenario for SQL Server on-premises and Azure SQL Database managed instance different to other RDBMS or even not possible in some cases:  

- Regular Azure SQL Database (single databases and elastic pools): NOT POSSIBLE
- Other RDBMS might NOT support the .NET CLR running inside the RDBMS server or have a particular implementation different to SQL Server CLR integration such as [Oracle Database Extensions for .NET](https://docs.oracle.com/cd/B19306_01/win.102/b14306/intro.htm) or PostgreSQL extension.

My point is that the 'ML.NET integration with a SQL-CLR approach' should probably be a different component with a different architecture/design and a different implementation per RDBMS, hence I believe we should have a different specs docs and different priorities depending on usage (i.e. CLR integration is not something very common in .NET applications, even less for .NET apps using non SQL Server RDBMS).But certainly, I find it particularly useful for intense operations such as training a model. :)

Reply:
However, I'm thinking we might be able to support 'CLR integration' at the same time if we are not really creating any database object but simply supporting to use the CLR within the database server.

The only difference is the way you define the connection string, such as:

//SQL Server
SqlConnection conn   = new SqlConnection("context connection=true")

//Oracle
OracleConnection con = new OracleConnection();
con.ConnectionString = "context connection=true"; 

So we can probably support both scenarios. 👍 

What would be a different implementation is if we were building concrete database objects such as CLR Scalar-Valued Functions or stored procedures which are different for SQL Server, Oracle, etc.

So, yeah, I think we can also support 'CLR integration' at the same time. 👍